refactor: enable `readability-container-contains` clang-tidy rule

l0rinc opened the pull request:
Replace the last few instances of `.count() != 0` and `.count() == 0` and bare `count()` patterns with the more expressive C++20 `.contains()` method:

* `std::set<std::string>` in `getblocktemplate` RPC;
* `std::map<std::string, ...>` in `transaction_tests`;
* other bare `std::unordered_set` and `std::map` count calls.

Also fixes https://github.com/bitcoin/bitcoin/issues/34101 by reverting `boost::multi_index::contains` calls not available in our minimum supported version.

With no remaining violations, enable the `readability-container-contains` clang-tidy check to prevent future regressions.

Follow-up to https://github.com/bitcoin/bitcoin/pull/33192

maflcko commented:
> Task `lint`: https://github.com/bitcoin/bitcoin/actions/runs/20319772315/job/58372602822

Looks like GitHub nuked all historic commits and historic CI runs here?

rkrux reviewed (approved):
lgtm, code review ACK 31c8ef3d9c885fbe810d5454ba1041ee87efa917

Nice, I missed reviewing the predecessor PR.

marcofleon reviewed (commented):
code review ACK a8a0a0ff77309a95645f6de00595c1dc39cd6eff

Tested with boost 1.74, good to go.

pablomartin4btc reviewed (approved):
cr-ACK a8a0a0ff77309a95645f6de00595c1dc39cd6eff

Verified the revert of `boost::multi_index::contains` instances (2).

janb84 reviewed (commented):
cr ACK a8a0a0ff77309a95645f6de00595c1dc39cd6eff

Also confirmed that clang-tidy (still) works

hebasto reviewed (commented):
Tested a8a0a0ff77309a95645f6de00595c1dc39cd6eff on Fedora 43 with the downloaded [Boost 1.73.0](https://www.boost.org/releases/1.73.0/) and the following patch:
```diff
--- a/cmake/module/AddBoostIfNeeded.cmake
+++ b/cmake/module/AddBoostIfNeeded.cmake
@@ -29,7 +29,7 @@ function(add_boost_if_needed)
     endif()
   endif()
 
-  find_package(Boost 1.73.0 REQUIRED CONFIG)
+  find_package(Boost 1.73.0 REQUIRED)
   mark_as_advanced(Boost_INCLUDE_DIR boost_headers_DIR)
   # Workaround for a bug in NetBSD pkgsrc.
   # See: https://github.com/NetBSD/pkgsrc/issues/167.
--- a/src/ipc/CMakeLists.txt
+++ b/src/ipc/CMakeLists.txt
@@ -20,6 +20,7 @@ target_link_libraries(bitcoin_ipc
   PRIVATE
     core_interface
     univalue
+    Boost::headers
 )
 
 if(BUILD_TESTS)
```

The compilation fails for another similar reason:
```
$ cmake -B build -DBoost_INCLUDE_DIR=/home/hebasto/Downloads/boost_1_73_0
$ cmake --build build
[398/621] Building CXX object src/CMakeFiles/bitcoin_node.dir/txmempool.cpp.o
FAILED: [code=1] src/CMakeFiles/bitcoin_node.dir/txmempool.cpp.o 
/usr/bin/ccache /usr/lib64/ccache/c++ -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE -I/home/hebasto/dev/bitcoin/build/src -I/home/hebasto/dev/bitcoin/src -I/home/hebasto/dev/bitcoin/src/leveldb/include -I/home/hebasto/dev/bitcoin/src/minisketch/include -I/home/hebasto/dev/bitcoin/src/univalue/include -isystem /home/hebasto/Downloads/boost_1_73_0 -O2 -g -std=c++20 -fPIC -fno-extended-identifiers -fdebug-prefix-map=/home/hebasto/dev/bitcoin/src=. -fmacro-prefix-map=/home/hebasto/dev/bitcoin/src=. -fstack-reuse=none -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wformat -Wformat-security -Wvla -Wredundant-decls -Wdate-time -Wduplicated-branches -Wduplicated-cond -Wlogical-op -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wunreachable-code -Wbidi-chars=any -Wundef -Wleading-whitespace=spaces -Wtrailing-whitespace=any -Wno-unused-parameter -MD -MT src/CMakeFiles/bitcoin_node.dir/txmempool.cpp.o -MF src/CMakeFiles/bitcoin_node.dir/txmempool.cpp.o.d -o src/CMakeFiles/bitcoin_node.dir/txmempool.cpp.o -c /home/hebasto/dev/bitcoin/src/txmempool.cpp 
/home/hebasto/dev/bitcoin/src/txmempool.cpp: In member function ‘void CTxMemPool::Apply(ChangeSet*)’:
/home/hebasto/dev/bitcoin/src/txmempool.cpp:205:48: error: ‘CTxMemPool::indexed_transaction_set’ {aka ‘class boost::multi_index::multi_index_container<CTxMemPoolEntry, CTxMemPool::CTxMemPoolEntry_Indices>’} has no member named ‘extract’
  205 |         auto node_handle = changeset->m_to_add.extract(tx_entry);
      |                                                ^~~~~~~
[415/621] Building CXX object src/CMakeFiles/bitcoin_node.dir/validation.cpp.o
ninja: build stopped: subcommand failed.
```

UPDATE: Boost 1.74 compiles without the error.

l0rinc commented:
Thanks @hebasto, I think the two can be merged independently.

hebasto reviewed (commented):
Approach ACK a8a0a0ff77309a95645f6de00595c1dc39cd6eff.

> Also fixes #34101 by reverting `boost::multi_index::contains` calls not available in our minimum supported version.

I'd prefer having this change in a separate commit.

Also there's a typo in the commit message:
"... available in Boost 1.74.0 only" --> "... available in Boost since 1.78.0 only"

l0rinc committed:
Fix compilation for old Boost versions

Fixes https://github.com/bitcoin/bitcoin/issues/34101 by reverting `boost::multi_index::contains` calls only available in Boost 1.78.0

l0rinc committed:
refactor: enable `readability-container-contains` clang-tidy rule

Replace the last few instances of `.count() != 0` and `.count() == 0` and `.count()` patterns with the more expressive C++20 `.contains()` method:

* `std::set<std::string>` in `getblocktemplate` RPC;
* `std::map<std::string, ...>` in `transaction_tests`;
* other bare `std::unordered_set` and `std::map` count calls.

With no remaining violations, enable the `readability-container-contains`
clang-tidy check to prevent future regressions.

hebasto reviewed (approved):
ACK 1e94e562f76e6152dffb2a2d07dc3429137098b5.

pablomartin4btc reviewed (approved):
re-ACK 1e94e562f76e6152dffb2a2d07dc3429137098b5

Boost fix/ revert split in a separate commit.

janb84 reviewed (commented):
ACK 1e94e562f76e6152dffb2a2d07dc3429137098b5

changes since last ack:
- reverted boost library contains calls
- split out the revert in separate commit.

rkrux reviewed (commented):
re-ACK 1e94e562f76e6152dffb2a2d07dc3429137098b5
